model/riscv_insts_zfh.sail

```diff
@@ -0,0 +1,934 @@
+      if   (e1 == e2)
+      then unsigned (m1) <=  unsigned (m2)
+      else unsigned (e1)  <  unsigned (e2)
+    else if (s1 == 0b0) & (s2 == 0b1)
+    then (v1Is0 & v2Is0)                         /* Equal in this case (+0=-0) */
+    else if (s1 == 0b1) & (s2 == 0b0)
+    then true
+    else
+      if   (e1 == e2)
+      then unsigned (m1) >=  unsigned (m2)
+      else unsigned (e1)  >  unsigned (e2);
+
+  let fflags = if is_quiet then
+                 if   (f_is_SNaN_H(v1) | f_is_SNaN_H(v2))
+                 then nvFlag()
+                 else zeros()
+               else
+                 if   (f_is_NaN_H(v1) | f_is_NaN_H(v2))
+                 then nvFlag()
+                 else zeros();
+
+  (result, fflags)
+}
+
+/* ****************************************************************** */
+/* Floating-point loads                                               */
+/* These are defined in: riscv_insts_fext.sail                        */
+
+/* ****************************************************************** */
+/* Floating-point stores                                              */
+/* These are defined in: riscv_insts_fext.sail                        */
+
+/* Binary ops with rounding mode */
+
+/* AST */
+
+union clause ast = F_BIN_RM_TYPE_H : (regidx, regidx, rounding_mode, regidx, f_bin_rm_op_H)
+
+/* AST <-> Binary encoding ================================ */
+
+mapping clause encdec =
+    F_BIN_RM_TYPE_H(rs2, rs1, rm, rd, FADD_H)                             if haveZfh()
+<-> 0b000_0010 @ rs2 @ rs1 @ encdec_rounding_mode (rm) @ rd @ 0b101_0011  if haveZfh()
+
+mapping clause encdec =
+    F_BIN_RM_TYPE_H(rs2, rs1, rm, rd, FSUB_H)                             if haveZfh()
+<-> 0b000_0110 @ rs2 @ rs1 @ encdec_rounding_mode (rm) @ rd @ 0b101_0011  if haveZfh()
+
+mapping clause encdec =
+    F_BIN_RM_TYPE_H(rs2, rs1, rm, rd, FMUL_H)                             if haveZfh()
+<-> 0b000_1010 @ rs2 @ rs1 @ encdec_rounding_mode (rm) @ rd @ 0b101_0011  if haveZfh()
+
+mapping clause encdec =
+    F_BIN_RM_TYPE_H(rs2, rs1, rm, rd, FDIV_H)                             if haveZfh()
+<-> 0b000_1110 @ rs2 @ rs1 @ encdec_rounding_mode (rm) @ rd @ 0b101_0011  if haveZfh()
+
+/* Execution semantics ================================ */
+
+function clause execute (F_BIN_RM_TYPE_H(rs2, rs1, rm, rd, op)) = {
+  let rs1_val_16b = nan_unbox_H (F(rs1));
+  let rs2_val_16b = nan_unbox_H (F(rs2));
+  match (select_instr_or_fcsr_rm (rm)) {
+    None() => { handle_illegal(); RETIRE_FAIL },
+    Some(rm') => {
+      let rm_3b = encdec_rounding_mode(rm');
+      let (fflags, rd_val_16b) : (bits(5), bits(16)) = match op {
+        FADD_H  => riscv_f16Add (rm_3b, rs1_val_16b, rs2_val_16b),
+        FSUB_H  => riscv_f16Sub (rm_3b, rs1_val_16b, rs2_val_16b),
+        FMUL_H  => riscv_f16Mul (rm_3b, rs1_val_16b, rs2_val_16b),
+        FDIV_H  => riscv_f16Div (rm_3b, rs1_val_16b, rs2_val_16b)
+      };
+      write_fflags(fflags);
+      F(rd) = nan_box_H (rd_val_16b);
```

Comment:
These can all be nan_box, for consistency with F (that or those should be changed to nan_box_S, but I don't see the benefit from that, might as well make use of the language's overloading to make the code a bit cleaner to read), though given they'll all need changing in future to `F_or_X_H(rd) = ...` with the NaN boxing implicit I guess this is fine, unlike the one for FP_LOAD which will stay and thus should be fixed (though arguably pushing that into a new F_H/S/D might make things slightly cleaner and more consistent, as a future minor refactor).